Upgrade CF components to newer versions

Proplex opened the pull request:
# Overview
This release updates CF components to more recent versions. Most
notably, this version brings the following related CF changes:
* Diego now uses mTLS for communication.
* CAPI now supports IAM Service Accounts when authenticating with
  Google Cloud Storage
* Scalable Syslog is now used, allowing multiple VMs to handle log
  output.


To accommodate these changes, the following kit-related changes were
made:
* 3 new databases are necessary: `silkdb`, `locketdb`, and 
  `policyserverdb`
* Parameters for `syslogger` VM type and instance count
* Self-signed certs for internal mTLS communication 


Various configuration parameters were added, and are detailed below.

# New Parameters

## Scalable Syslog

  - `syslogger_instances` - How many scalable syslog VMs to deploy.

  - `syslogger_vm_type` - What type of VM to deploy for the scalable
    syslog. Defaults to `small`.

## UAA

  - `uaa_lockout_failure_count` - Amount of failed UAA login attempts
    before lockout.

  - `uaa_lockout_failure_time_between_failures` - How much time
    (in seconds) in which `uaa_lockout_failure_count` must occur in 
    order for account to be locked. Defaults to `1200`.
  
  - `uaa_lockout_punishment_time` - How long (in seconds) the account
    is locked out for violating `uaa_lock_failure_count` within 
    `uaa_lockout_failure_time_between_failures`. Defaults to `300`.

  - `uaa_refresh_token_validity` - How long (in seconds) a CF refresh
    is valid for. Defaults to `2592000`.

  - `cf_branding_product_logo` - A base64 encoded image to display on
    the web UI login prompt. Defaults to `nil`.

  - `cf_branding_square_logo` - A base64 encoded image to display
    in areas where a smaller logo is necessary. Defaults to `nil`.

  - `cf_footer_legal_text` - A string to display in the footer,
    typically used for compliance text. Defaults to `nil`.

  - `cf_footer_links` - A YAML list of links to enumerate in the footer
    of the web UI. Defaults to `nil`

# Upgrade Instructions

From `1.0.0`, the following actions need to be made:

## If CF was deployed with `local-db` or `local-ha-db` feature flags

No changes need to be performed for an in-situ upgrade of 
`cf-genesis-kit` 1.0.0 to 1.1

## If CF was deployed with `mysql-db` or `postgres-db` feature flags

1. Create 3 databases with the following names: `silkdb`, `locketdb`, and 
  `policyserverdb`.



# Core Components

| Release | Version | Release Date |
| --------- | ------- | ------------ |
bpm (new) | [0.6.0](https://github.com/cloudfoundry-incubator/bpm-release/releases/tag/v0.6.0) | 2 May 2018
capi | [1.60.0](https://github.com/cloudfoundry/capi-release/releases/tag/1.60.0) | 15 Jun 2018
cf-smoke-tests | 40.0.5 | 17 May 2018
cf-networking (new) | [1.9.0](https://github.com/cloudfoundry/cf-networking-release/releases/tag/v1.9.0) | 15 Dec 2017
cflinuxfs2 | [1.212.0](https://github.com/cloudfoundry/cflinuxfs2-release/releases/tag/v1.212.0) | 4 Jun 2018
cf-syslog-drain (new) | [6.5](https://github.com/cloudfoundry/cf-syslog-drain-release/releases/tag/v6.5) | 3 May 2018
consul | [193](https://github.com/cloudfoundry-incubator/consul-release/releases/tag/v193) | 29 May 2018
diego | [2.8.0](https://github.com/cloudfoundry/diego-release/releases/tag/v2.8.0) | 28 May 2018
garden-runc | [1.14.0](https://github.com/cloudfoundry/garden-runc-release/releases/tag/v1.14.0) | 4 Jun 2018
loggregator | [102.2](https://github.com/cloudfoundry/loggregator-release/releases/tag/v102.2) | 25 May 2018
nats | [24](https://github.com/cloudfoundry/nats-release/releases/tag/v24) | 16 May 2018 
cf-routing | [0.178.0](https://github.com/cloudfoundry/routing-release/releases/tag/0.178.0) | 17 May 2018
statsd-injector | [1.3.0](https://github.com/cloudfoundry/statsd-injector-release/releases/tag/v1.3.0) | 23 Mar 2018
uaa | [59](https://github.com/cloudfoundry/uaa-release/releases/tag/v59) | 22 May 2018


# Buildpacks

| Buildpack | Version | Release Date |
| --------- | ------- | ------------ |
binary | [1.0.19](https://github.com/cloudfoundry/binary-buildpack/releases/tag/v1.0.19) | 5 Jun 2018
dotnet-core | [2.0.7](https://github.com/cloudfoundry/dotnet-core-buildpack/releases/tag/v2.0.7) | 5 Jun 2018
go | [1.8.23](https://github.com/cloudfoundry/go-buildpack/releases/tag/v1.8.23) | 5 Jun 2018
java | [4.12](https://github.com/cloudfoundry/java-buildpack/releases/tag/v4.12) | 11 May 2018
nodejs | [1.6.25](https://github.com/cloudfoundry/nodejs-buildpack/releases/tag/v1.6.25) | 5 Jun 2018
php | [4.3.56](https://github.com/cloudfoundry/php-buildpack/releases/tag/v4.3.56) | 5 Jun 2018
python | [1.6.17](https://github.com/cloudfoundry/python-buildpack/releases/tag/v1.6.17) | 5 Jun 2018
ruby | [1.7.19](https://github.com/cloudfoundry/ruby-buildpack/releases/tag/v1.7.19) | 5 Jun 2018
staticfile | [1.4.28](https://github.com/cloudfoundry/staticfile-buildpack/releases/tag/v1.4.28) | 5 Jun 2018

Proplex commented:
This upgrade is still being tested in GCP. I'd like to perform more testing before pulling this into master, but I would still appreciate feedback in the meantime.

jhunt commented on MANUAL.md:
```diff
@@ -29,6 +29,36 @@ Cloud Foundry.
+    requirements, explicitly. Defaults to `1000`.
+
+  - `uaa_lockout_failure_count` - Amount of failed UAA login attempts
+    before lockout.
```
"Number" works better here, since "amount" has connotation of being large (amount of money) or even uncountable (amount of good will).

jhunt commented on MANUAL.md:
```diff
@@ -29,6 +29,36 @@ Cloud Foundry.
+
+  - `uaa_lockout_failure_time_between_failures` - How much time
+    (in seconds) in which `uaa_lockout_failure_count` must occur in 
+    order for account to be locked. Defaults to `1200`.
```
This name is a bit unwieldy.

Perhaps `uaa_lockout_window`?

jhunt commented on MANUAL.md:
```diff
@@ -29,6 +29,36 @@ Cloud Foundry.
+  
+  - `uaa_lockout_punishment_time` - How long (in seconds) the account
+    is locked out for violating `uaa_lock_failure_count` within 
+    `uaa_lockout_failure_time_between_failures`. Defaults to `300`.
```
Punishment seems a bit harsh.

Perhaps `uaa_lockout_time`?

patrick1378 replied:
Also the parameter is `uaa_lockout_time_between_failures`, but if were changing the name anyways.

jhunt commented on MANUAL.md:
```diff
@@ -29,6 +29,36 @@ Cloud Foundry.
+    order for account to be locked. Defaults to `1200`.
+  
+  - `uaa_lockout_punishment_time` - How long (in seconds) the account
+    is locked out for violating `uaa_lock_failure_count` within 
```
The referenced param is actually named `uaa_lockout_failure_count`

jhunt commented on MANUAL.md:
```diff
@@ -29,6 +29,36 @@ Cloud Foundry.
+    typically used for compliance text. Defaults to `nil`.
+
+  - `cf_footer_links` - A YAML list of links to enumerate in the footer
+    of the web UI. Defaults to `nil`
```
How are the links specified?  As raw HTML?  As an object containing link text and URL?  How would an operator use this?

jhunt commented on MANUAL.md:
```diff
@@ -29,6 +29,36 @@ Cloud Foundry.
+    is valid for. Defaults to `2592000`.
+
+  - `cf_branding_product_logo` - A base64 encoded image to display on
+    the web UI login prompt. Defaults to `nil`.
```
Can we write up a small thing on branding (in a different section of MANUAL.md) and point to it in these param defs?  Could be a good place to provide some examples, explain why base64 is required, etc.

jhunt commented on ci/release_notes.md:
```diff
@@ -1,10 +1,107 @@
+* CAPI now supports IAM Service Accounts when authenticating with
+  Google Cloud Storage
+* Scalable Syslog is now used, allowing multiple VMs to handle log
+  output.
```
Separate the paragraph from the list below it and the header above it, with blank lines.

Also, if you can convert the lists to use `-` instead of `*` for the bullet sigil, that would be more in-line with all of our other release notes.

jhunt commented on ci/release_notes.md:
```diff
@@ -1,10 +1,107 @@
+made:
+* 3 new databases are necessary: `silkdb`, `locketdb`, and 
+  `policyserverdb`
+* Parameters for `syslogger` VM type and instance count
```
These have defaults and therefore do not need to be listed in the release notes.

jhunt commented on ci/release_notes.md:
```diff
@@ -1,10 +1,107 @@
+* 3 new databases are necessary: `silkdb`, `locketdb`, and 
+  `policyserverdb`
+* Parameters for `syslogger` VM type and instance count
+* Self-signed certs for internal mTLS communication 
```
These are hands-off (I believe) and do not require operator involvement, so they don't need to be in the release notes.

Related: do we need to mention the SAN changes and how those will be handled?

jhunt commented on ci/release_notes.md:
```diff
@@ -1,10 +1,107 @@
+
+Various configuration parameters were added, and are detailed below.
+
+# New Parameters
```
This section might be a bit much, you can probably skip the detailed descriptions and just let the operators know that there are now things, and what section of MANUAL.md to refer to.

jhunt commented on ci/release_notes.md:
```diff
@@ -1,10 +1,107 @@
+nats | [24](https://github.com/cloudfoundry/nats-release/releases/tag/v24) | 16 May 2018 
+cf-routing | [0.178.0](https://github.com/cloudfoundry/routing-release/releases/tag/0.178.0) | 17 May 2018
+statsd-injector | [1.3.0](https://github.com/cloudfoundry/statsd-injector-release/releases/tag/v1.3.0) | 23 Mar 2018
+uaa | [59](https://github.com/cloudfoundry/uaa-release/releases/tag/v59) | 22 May 2018
```
Question (which may/should engender a new ticket):

Can/are we programmatically extracting this information from the kit yamls?

Proplex replied:
Yes, please. That md list was the most tedious part of this entire PR. I can write a script that automatically generates something like this for future releases.

jhunt commented on hooks/pre-deploy:
```diff
@@ -0,0 +1,11 @@
+
+#An in-situ upgrade of CF required $GENESIS_VAULT_PREFIX/diego/certs/rep to include 127.0.0.1
+#as a SAN. Rather than bothering the operator for an internal cert with no impact on external
+#performance, it's better to just delete and regen the cert automatically.
```
Nit: always put spaces between your comment character and the comment text.

jhunt commented on hooks/pre-deploy:
```diff
@@ -0,0 +1,11 @@
+  safe rm "$vault/diego/certs/rep"
+  vault_target=$(safe target --json | jq --raw-output '.name')
+  genesis -C $GENESIS_ROOT add-secrets $GENESIS_ENVIRONMENT --vault $vault_target >/dev/null 2>&1
+fi
```
Can we do this holistically, for all certs?

If it works well, we may migrate that behavior into genesis proper.  But I'd like to first experiment with this kit.

jhunt commented on manifests/cf/cell.yml:
```diff
@@ -9,35 +9,46 @@ instance_groups:
         persistent_image_list:
         - "/var/vcap/packages/cflinuxfs2/rootfs.tar"
+      grootfs:
+        reserved_space_for_other_jobs_in_mb: 15360
```
Does this need to be configurable?

jhunt commented on manifests/cf/certs.yml:
```diff
@@ -104,8 +126,22 @@ meta:
+
     uaa:
-      ca: (( vault meta.vault "/uaa/certs/ca:certificate" ))
+      ca: (( vault meta.vault "/uaa/certs/ca:certificate" )) 
```
WHITESPACE!

jhunt commented on manifests/cf/cloud_controller.yml:
```diff
@@ -13,6 +13,9 @@ instance_groups:
             cloud_controller_ng: {}
+            policy-server-internal:
+              name: policy-server
+
```
Might want to include a comment here as to why the name and the key differ (iirc, it's because of the health check path)

jhunt commented on manifests/cf/diego.yml:
```diff
@@ -13,6 +13,11 @@ instance_groups:
+          use_v2_api: true
+          ca_cert: (( grab meta.certs.loggregator.ca ))
+          cert: (( grab meta.certs.loggregator.metron.server.cert ))
+          key: (( grab meta.certs.loggregator.metron.server.key ))
```
I've seen this block a bunch; does it make sense to meta it and just `(( grab meta.loggregator.config ))`?

Proplex replied:
I'll have this sorted out for #24

jhunt commented on manifests/cf/uaa.yml:
```diff
@@ -155,17 +177,30 @@ instance_groups:
+    product_logo: (( grab params.cf_branding_product_logo || nil ))
+    square_logo: (( grab params.cf_branding_square_logo  || nil ))
+    footer_legal_text: (( grab params.cf_branding_footer_legal_text || nil ))
+    footer_links: (( grab params.cf_branding_footer_links || nil ))
```
Nit: can you line up the opening `((`s?

jhunt commented on manifests/cf/uaa.yml:
```diff
@@ -155,17 +177,30 @@ instance_groups:
 
+    failure_count: (( grab params.uaa_lockout_failure_count || 5 ))
+    failure_timeout: (( grab params.uaa_lockout_punishment_time || 300 ))
+    time_between_failures: (( grab params.uaa_lockout_time_between_failures || 1200 ))
```
Nit: line up the `((`s, and the `||`s -- spruce ignores interior whitespace

patrick1378 commented on kit.yml:
```diff
@@ -229,6 +271,10 @@ credentials:
+      password: random 64
 
   autoscaler:
     autoscaler/servicebroker_account:
```
Missing generating password for the service broker account.

Proplex replied:
Taken care of on L282

Proplex commented:
I've addressed all the feedback so far. I've also written a small document explaining each VM's purpose and what each job does. It's located here atm: https://github.com/Proplex/cheatsheets/blob/master/CF%20Genesis%20Kit%20Explanation.md not sure if it's something we should roll into this repo.

Proplex commented on manifests/cf/cell.yml:
```diff
@@ -9,35 +9,46 @@ instance_groups:
         persistent_image_list:
         - "/var/vcap/packages/cflinuxfs2/rootfs.tar"
+      grootfs:
+        reserved_space_for_other_jobs_in_mb: (( grab params.grootfs_reserved_space || 15360 ))
```
Is there a method to retrieve the disk space of the VM within `.yml`?

dennisjbell replied:
nope, you'd have to create a file in the kit directory during the blueprint hook and populate it, then list that file as a merge candidate.  However, I'm not sure we need to do that, as we can just specify a value in params  -- I was just wondering what WOULD happen if the reserve is bigger than the disk size (that's why it wasn't a required change, unlike the output issue listed above)

Proplex commented:
Blueprint output issue corrected, and I believe that the `reserved_space_for_other_jobs_in_mb` is OK to keep. See comments above.
